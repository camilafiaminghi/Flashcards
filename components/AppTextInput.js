import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, View, Text, TextInput } from 'react-native'
import { validationRules } from '../utils/validation'
import { gray, black } from '../utils/colors'

class AppTextInput extends Component {

	static propTypes = {
		name: PropTypes.string.isRequired,
		placeholder: PropTypes.string.isRequired,
		maxLen: PropTypes.number.isRequired,
		onInputChange: PropTypes.func
	}

	state = {
		text: '',
		changed: false,
		valid: false
	}

	handleValidation = (value) => {
		const { name, onInputChange, maxLen } = this.props
		const valid = validationRules(name, value)

		if ( value.length <= maxLen ) {
			this.setState((state) => ({
				...state,
				text: value,
				changed: true,
				valid
			}))

			onInputChange(name, value, valid)
		}
	}

	render() {
		const { text, changed, valid } = this.state
		const { maxLen, placeholder } = this.props
		const textLeft = maxLen - text.length

		return (
			<View style={styles.container}>
				<TextInput
	        style={[styles.input, styles.inputText]}
	        placeholder={placeholder}
	        onChangeText={this.handleValidation}
	        value={text} />

	      { (textLeft !== 0 && textLeft < maxLen && changed) && <Text style={styles.text}>{ textLeft } characteres left</Text> }
	      { (textLeft === 0 && changed) && <Text style={styles.text}>max characters exceeded</Text>}
			</View>
		)
	}
}

export default AppTextInput

/* STYLES */
const styles = StyleSheet.create({
	container: {
		alignSelf: 'stretch',
		margin: 10
	},
	input: {
		flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderColor: gray,
		borderWidth: 1,
		borderRadius: 4
	},
	inputText: {
		padding: 8
	},
	text: {
		fontSize: 12,
		color: black
	}
})
