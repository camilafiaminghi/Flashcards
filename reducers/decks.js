import {
		ADD_DECK,
		ADD_CARD } from '../actions'

const decks = (state = {}, action) => {
	switch (action.type) {
		case ADD_DECK:
			return {
				...state,
				[action.payload.title] : {
					title: action.payload.title,
					questions: []
				}
			}
		case ADD_CARD:
			return {
				...state,
				[action.payload.deckId] : {
					...state[action.payload.deckId],
					questions: [...state[action.payload.deckId].questions, action.payload.card]
				}
			}
		default:
			return state
	}
}

export default decks
