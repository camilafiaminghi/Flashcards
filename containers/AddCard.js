import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { NavigationActions } from 'react-navigation'
import { connect } from 'react-redux'
import { StyleSheet, KeyboardAvoidingView, Text, View, TouchableOpacity } from 'react-native'
import { handleAddCard } from '../actions'
import { textColorInverse, textColor, pColor, pColorLight, pColorDark, sColor, sColorLight, sColorDark } from '../utils/colors'
import { isValid } from '../utils/validation'
import AppTextInput from '../components/AppTextInput'

export class AddCard extends Component {

	static navigationOptions = ({ navigation }) => {
		return {
			title: 'New Card'
		}
	}

	static propTypes = {
		deckId: PropTypes.string.isRequired
	}

	state = {
		card: {
			question: '',
			answer: ''
		},
		validation: {
			question: false,
			answer: false
		},
		valid: false
	}

	onInputChange = (name, value, valid) => {
		this.setState((state) => ({
			...state,
			card: {
				...state.card,
				[name]: value
			},
			validation: {
				...state.validation,
				[name]: valid
			}
		}))

		/* VALIDATE ALL */
		this.setState((state) => ({
			...state,
			valid: isValid(state.validation)
		}))
	}

	toDeck = () => {
		const { deckId } = this.props

		this.props.navigation.dispatch(NavigationActions.navigate({
			routeName: 'Deck',
			params: { deckId }
		}))
	}

	submit = () => {
		const { onAddCard, deckId } = this.props
		const { card } = this.state

		onAddCard(deckId, card)
			.then(this.toDeck) /* Navigate to Deck */
	}

	render() {
		const { valid } = this.state

		return (
			<View style={styles.container}>
				<KeyboardAvoidingView behavior='padding' style={styles.card}>

					<AppTextInput
						name='question'
						placeholder='Question? (max 122 characters)'
						maxLen={122}
						onInputChange={this.onInputChange} />

					<AppTextInput
						name='answer'
						placeholder='Answer (max 255 characters)'
						maxLen={255}
						onInputChange={this.onInputChange} />

				  <TouchableOpacity
		      	style={[styles.input, styles.btn]}
		      	disabled={!valid}
		      	onPress={this.submit}>
		      	<Text style={ valid ? styles.btnText : styles.btnDisabledText }>{'submit'.toUpperCase()}</Text>
		      </TouchableOpacity>
				</KeyboardAvoidingView>
			</View>
		)
	}
}

export const mapStateToProps = (state, { navigation }) => {
	const { deckId } = navigation.state.params
	return {
		deckId
	}
}

export const mapDispatchToProps = (dispatch) => {
	return {
		onAddCard: (deckId, card) => dispatch(handleAddCard(deckId, card))
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(AddCard)

/* STYLES */
const styles = StyleSheet.create({
	container: {
    flex: 1,
    backgroundColor: sColorLight,
    paddingTop: 4,
    paddingBottom: 4
  },
  card: {
  	flex: 1,
  	flexGrow: 1,
  	alignItems: 'center',
  	justifyContent: 'center',
  	backgroundColor: 'white',
  	margin: 8,
		padding: 8,
  	borderRadius: 4,
  },
  text: {
  	flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderColor: textColor,
		borderWidth: 1,
		borderRadius: 4
  },
	input: {
		flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderBottomColor: pColorDark,
		borderBottomWidth: 1,
		borderRadius: 3,
	},
	btn: {
		alignItems: 'center',
		justifyContent: 'center',
		padding: 4,
		marginRight: 10,
		marginLeft: 10
	},
	btnDisabledText: {
		fontSize: 14,
		fontWeight: 'bold',
		color: pColorLight
	},
	btnText: {
		fontSize: 14,
		fontWeight: 'bold',
		color: pColorDark
	}
})
