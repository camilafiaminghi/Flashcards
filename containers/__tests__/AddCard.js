import React, { Component } from 'react'
import { TouchableOpacity } from 'react-native'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { AddCard, mapStateToProps, mapDispatchToProps } from '../AddCard'
import AppTextInput from '../../components/AppTextInput'
import decks from '../../__mocks__/decks'

let wrapper
let instance
let component
const navigation = {
	navigate: jest.fn(),
	state: {
		params: {
			deckId: Object.keys(decks)[0]
		}
	},
	getParam: jest.fn(),
	setParams: jest.fn(),
	dispatch: jest.fn().mockResolvedValue()
}
const props = {
	navigation,
	deckId: Object.keys(decks)[0],
	onAddCard: jest.fn().mockResolvedValue()
}

const configStore = configureMockStore([thunk])
const store = configStore({decks})

describe('<AddCard />', () => {

	beforeEach(() => {
		wrapper = renderer.create(<AddCard {...props} />)
		instance = shallow(<Provider store={store}><AddCard {...props} /></Provider>)
		component = instance.find(AddCard).shallow()
	})

	afterEach(() => store.clearActions())

	it('matches snapshot', () => {
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

	it('should render', () => {
		expect(wrapper).toBeTruthy()
	})

	it('should has two AppTextInput as children', () => {
		expect(component.find(AppTextInput)).toHaveLength(2)
	})

	it('should has submit if inputs are valid', () => {
		component.instance().onInputChange('question', 'Question Test', true)
		component.instance().onInputChange('answer', 'Answer Test', true)

		expect(component.state('valid')).toEqual(true)
		expect(component.find(TouchableOpacity)).toHaveLength(1)
	})

	it('should handle submit', () => {
		component.instance().onInputChange('question', 'Question Test', true)
		component.instance().onInputChange('answer', 'Answer Test', true)

		const spy = jest.spyOn(component.instance(), 'submit')
		const button = component.find(TouchableOpacity)

		button.props().onPress()
		expect(props.onAddCard).toHaveBeenCalledTimes(1)
		expect(props.onAddCard).toHaveBeenCalledWith(props.deckId, component.state('card'))
	})

	it('should mapStateToProps return props', () => {
		expect(mapStateToProps(store.getState(), { navigation })).toHaveProperty('deckId')
	})

	it('should mapDispatchToProps return props', () => {
		expect(mapDispatchToProps(store.dispatch)).toHaveProperty('onAddCard')
	})
})
