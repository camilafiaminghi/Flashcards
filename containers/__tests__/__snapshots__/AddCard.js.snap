// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCard /> matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#efefef",
      "flex": 1,
      "paddingBottom": 4,
      "paddingTop": 4,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 4,
          "flex": 1,
          "flexGrow": 1,
          "justifyContent": "center",
          "margin": 8,
          "padding": 8,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "margin": 10,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Question? (max 122 characters)"
        placeholderTextColor="#6f79a8"
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderBottomColor": "#6f79a8",
              "borderBottomWidth": 2,
              "borderRadius": 3,
              "flexDirection": "row",
              "height": 50,
            },
            Object {
              "padding": 8,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "margin": 10,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Answer (max 255 characters)"
        placeholderTextColor="#6f79a8"
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderBottomColor": "#6f79a8",
              "borderBottomWidth": 2,
              "borderRadius": 3,
              "flexDirection": "row",
              "height": 50,
            },
            Object {
              "padding": 8,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "borderBottomColor": "#6f79a8",
          "borderBottomWidth": 1,
          "borderRadius": 3,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "padding": 4,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#d1d9ff",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        SUBMIT
      </Text>
    </View>
  </View>
</View>
`;
